legacy/signals/signals.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package signals
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+// sigMap holds all signals the Watch() function should terminate
+// when encountering. The key holds the string representation
+// of the captured os.Signal. The value is the actual syscall of the
+// signal which allows the Watch() function to terminate with that exit code.
+var sigMap = map[string]syscall.Signal{
+	"hangup":              syscall.SIGHUP,
+	"interrupt":           syscall.SIGINT,
+	"aborted":             syscall.SIGABRT,
+	"illegal instruction": syscall.SIGILL,
+	"quit":                syscall.SIGQUIT,
+	"terminated":          syscall.SIGTERM,
+	"segmentation fault":  syscall.SIGSEGV,
+	"stopped":             syscall.SIGTSTP,
+}
+
+// Watch concurrenlty logs debug statements when encountering interrupt signals from the OS.
+// This function relies on the os/signal package which may not capture SIGKILL and SIGSTOP.
+func Watch() {
+	c := make(chan os.Signal, 1)
+	// Observe all signals, excluding SIGKILL and SIGSTOP.
+	signal.Notify(c)
+	// Continuously watch for signals.
+	go func() {
+		logrus.Debug("Watching for OS Signals...")
+		for sig := range c {
+			logrus.Debug("Encoutered signal: ", sig.String())
+			// Ensure the current program stops when encountering
+			// an interrupt. Without this check, these signals would
+			// not cause the program to terminate.
+			if s, exists := sigMap[sig.String()]; exists {
+				os.Exit(int(s))
```

Comment:
It's a bit odd to exit with the integer value of the signal. Better to log it (just include it in the Debug line above at line 52). And then here you can additionally log `"Exiting because of signal %s (%s)", sig.String(), sig.Signal()` or something similar.

Reply:
Concurrently capturing signals requires me to explicitly exit. The above comment explains this behavior. If I didn't do this, and a `SIGINT` was given, the program would not terminate. But I can add an extra logging statement to show the program is about to exit.

However, a [signal](https://pkg.go.dev/syscall#Signal) is of type int, so this casting is actually quite trivial.

Reply:
The `os.Exit()` is fine, but exiting with the `int(s)` as argument is what is awkward. Better to log it as `int(s)` or some such in the Debug line above it.